sdk/core/azure-core/src/main/java/com/azure/core/util/ConfigurationBuilder.java

```diff
@@ -0,0 +1,249 @@
+     * </pre>
+     * <!-- end com.com.azure.core.util.ConfigurationBuilder#addProperty -->
+     */
+    public ConfigurationBuilder() {
+        this.mutableSource = new MutableConfigurationSource();
+        this.environmentConfiguration = EnvironmentConfiguration.getGlobalConfiguration();
+    }
+
+    /**
+     * Creates {@code ConfigurationBuilder} with configuration source.
+     *
+     * <!-- src_embed com.azure.core.util.Configuration -->
+     * <pre>
+     * &#47;&#47; Creates Configuration with configured root path to shared properties
+     * Configuration configuration = new ConfigurationBuilder&#40;new SampleSource&#40;properties&#41;&#41;
+     *     .root&#40;&quot;azure.sdk&quot;&#41;
+     *     .buildSection&#40;&quot;client-name&quot;&#41;;
+     *
+     * ConfigurationProperty&lt;String&gt; proxyHostnameProperty = ConfigurationPropertyBuilder.ofString&#40;&quot;http.proxy.hostname&quot;&#41;
+     *     .shared&#40;true&#41;
+     *     .build&#40;&#41;;
+     * System.out.println&#40;configuration.get&#40;proxyHostnameProperty&#41;&#41;;
+     * </pre>
+     * <!-- end com.com.azure.core.util.Configuration -->
+     * @param source Custom {@link ConfigurationSource} containing known Azure SDK configuration properties.
+     */
+    public ConfigurationBuilder(ConfigurationSource source) {
+        this.mutableSource = new MutableConfigurationSource(Objects.requireNonNull(source, "'source' cannot be null"));
+        this.environmentConfiguration = EnvironmentConfiguration.getGlobalConfiguration();
+    }
+
+    /**
+     * Creates {@code ConfigurationBuilder} with configuration sources for explicit configuration, system properties and environment configuration sources.
+     * Use this constructor to customize known Azure SDK system properties and environment variables retrieval.
+     *
+     * @param source Custom {@link ConfigurationSource} containing known Azure SDK configuration properties
+     * @param systemPropertiesConfigurationSource {@link ConfigurationSource} containing known Azure SDK system properties.
+     * @param environmentConfigurationSource {@link ConfigurationSource} containing known Azure SDK environment variables.
+     */
+    public ConfigurationBuilder(ConfigurationSource source, ConfigurationSource systemPropertiesConfigurationSource, ConfigurationSource environmentConfigurationSource) {
+        Objects.requireNonNull(source, "'source' cannot be null");
+        Objects.requireNonNull(systemPropertiesConfigurationSource, "'systemPropertiesConfigurationSource' cannot be null");
+        Objects.requireNonNull(environmentConfigurationSource, "'environmentConfigurationSource' cannot be null");
+        this.mutableSource = new MutableConfigurationSource(source);
+        this.environmentConfiguration = new EnvironmentConfiguration(systemPropertiesConfigurationSource, environmentConfigurationSource);
+    }
+
+    /**
+     * Add property to the configuration source.
```

Comment:
We should also document that adding a property with the same name as an existing property will override the value (unlike the behavior in HttpHeaders). Also, if an instance of ConfigurationBuilder was created from source, the value coming from the source will be overwritten when addProperty is called.